File: 2025-04-04/Reverse_Linked_List_II.java
package amazon;

import java.util.List;

public class Reverse_Linked_List_II {
    public static class ListNode {
        int val;
        ListNode next;
        ListNode() {}
        ListNode(int val) { this.val = val; }
        ListNode(int val, ListNode next) { this.val = val; this.next = next; }
    }
    public static ListNode reverseBetween(ListNode head, int left, int right) {
        ListNode cur = head, prev = null;
        while(cur != null) {
            if(cur.val == left && left != right) {
                ListNode runPrev = cur, run = cur.next, last = cur;
                while(run != null) {
                    ListNode temp = run.next;
                    run.next = runPrev;

                    runPrev = run;
                    run = temp;

                    if(runPrev.val == right)
                        break;
                }
                if(prev != null) {
                    prev.next = runPrev;
                }
                else {
                    head = runPrev;
                }
                last.next = run;
                cur =run;
            }
            else {
                prev = cur;
                cur = cur.next;
            }
        }

        return head;
    }
    public static void main(String[] args) {
        ListNode head = new ListNode(1, new ListNode(2, new ListNode(3, new ListNode(4, new ListNode(5)))));
        ListNode ans = reverseBetween(head, 1, 5);

        while(ans!=null) {
            System.out.println(ans.val);
            ans = ans.next;
        }
    }
}

File: 2026-02-03/Reverse_Linked_List_II.py
from typing import Optional
from Linked_List import Linked_List, ListNode


class Solution:
    def reverseBetween(self, head: Optional[ListNode], left: int, right: int) -> Optional[ListNode]:
        copy = ListNode(0, head)
        prev = copy
        start = head
        ind = 1

        while start and ind != left:
            start = start.next
            prev = prev.next
            ind += 1

        revPrev = None
        cur = start

        while cur and ind <= right:
            temp = cur.next
            cur.next = revPrev
            revPrev = cur
            cur = temp
            ind += 1

        prev.next = revPrev
        start.next = cur
        return copy.next


sol = Solution()
ll = Linked_List()
head = ll.build_from_list([3,5])
ll.print_list(sol.reverseBetween(head, 1, 2))